packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,140 @@
+import { createAsyncMiddleware } from '@metamask/json-rpc-engine';
+import type {
+  NetworkClientId,
+  NetworkControllerFindNetworkClientIdByChainIdAction,
+  NetworkControllerGetNetworkClientByIdAction,
+  NetworkControllerGetStateAction,
+  NetworkControllerSetActiveNetworkAction,
+  NetworkControllerSetProviderTypeAction,
+} from '@metamask/network-controller';
+import type { SelectedNetworkControllerSetNetworkClientIdForDomainAction } from '@metamask/selected-network-controller';
+import { SelectedNetworkControllerActionTypes } from '@metamask/selected-network-controller';
+
+import type { QueuedRequestControllerEnqueueRequestAction } from './QueuedRequestController';
+import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
+
+const isConfirmationMethod = (method: string) => {
+  const confirmationMethods = [
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return !confirmationMethods.includes(method);
+};
+
+export const createQueuedRequestMiddleware = (
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<any, any> => {
+  return createAsyncMiddleware(async (req: any, res: any, next) => {
+    if (!useRequestQueue() || isConfirmationMethod(req.method)) {
+      next();
+      return;
+    }
+
+    const networkClientIdForRequest = req.networkClientId as NetworkClientId;
+
+    await messenger.call(
+      QueuedRequestControllerActionTypes.enqueueRequest,
+      async () => {
+        if (req.method === 'wallet_switchEthereumChain') {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        const isBuiltIn = isNetworkType(networkClientIdForRequest);
+        let networkConfigurationForRequest;
+        if (isBuiltIn) {
+          const builtIn = BUILT_IN_NETWORKS[networkClientIdForRequest];
+          if (builtIn.chainId) {
+            networkConfigurationForRequest = builtIn;
+          }
+        }
+
+        networkConfigurationForRequest =
+          networkConfigurationForRequest ||
+          messenger.call(
+            'NetworkController:getNetworkClientById',
+            networkClientIdForRequest,
+          ).configuration;
+
+        const currentProviderConfig = messenger.call(
+          'NetworkController:getState',
+        ).providerConfig;
+
+        const currentChainId = currentProviderConfig.chainId;
+
+        if (currentChainId === networkConfigurationForRequest.chainId) {
```

Comment:
Are we checking here that the globally selected network is the same as the dapp's selected network? It might be good to add a comment to this, or to extract this to a well-named function.